STS Storage Location Docs (#713)
Sage-Bionetworks/synapseDocs: _articles/aws_sts_storage_locations.md

## Synapse-Managed STS Storage Locations

You can create an STS Storage Location using Synapse storage. Synapse will store the files for you, but you can still get temporary S3 credentials, scoped to just the files and folders within your STS Storage Location.

To set up the STS Storage Location using Synapse storage, first make sure you have an empty Synapse folder, then run the following code:

##### Python

```python
# Set storage location
import synapseclient
import json
syn = synapseclient.login()
FOLDER = 'syn12345'

destination = {'uploadType':'S3',
               'stsEnabled':True,
               'concreteType':'org.sagebionetworks.repo.model.project.S3StorageLocationSetting'}
destination = syn.restPOST('/storageLocation', body=json.dumps(destination))

project_destination ={'concreteType': 'org.sagebionetworks.repo.model.project.UploadDestinationListSetting',
                      'settingsType': 'upload'}
project_destination['locations'] = [destination['storageLocationId']]
project_destination['projectId'] = FOLDER

> project_destination = syn.restPOST('/projectSettings', body = json.dumps(project_destination))

Reviewer: What is the name of the bucket created?  You will need that to download files via s3

Reviewer 2: Is it returned in the POST call to `/storageLocation`? Looks like a BucketStorageLocationSetting has the bucket name but I'm not sure if that's the implementation returned in this case.

Reviewer: It's not. There is a get `/storageLocation` but currently the s3 bucket name isn't exposed

Reviewer: Edit:  There is a way to get it by just doing a `syn.get(..., downloadFile=False)`

Author: I believe (but need to confirm for sure) that you can only get the File Handle back if you're the File Handle owner. Which means it will work for your own File Handles, but not for File Handles that are shared with you.

I've filed https://sagebionetworks.jira.com/browse/PLFM-6226 to make this improvement.

Reviewer 3: @DwayneJengSage Can you confirm that the destination['baseKey'] is the root of the scope of the bucket to which the STS token allows access?

Author: Yes, baseKey is the root folder to which the STS token allows access. It's an S3 prefix, so it doesn't contain the bucket name.

```

##### R

```r
#set storage location
library(synapser)
library(rjson)
synLogin()
folderId <- 'syn12345'

destination <- list(uploadType='S3',
                    stsEnabled=TRUE,
                    concreteType='org.sagebionetworks.repo.model.project.S3StorageLocationSetting')
destination <- synRestPOST('/storageLocation', body=toJSON(destination))

projectDestination <- list(concreteType='org.sagebionetworks.repo.model.project.UploadDestinationListSetting',
                           settingsType='upload')
projectDestination$locations <- list(destination$storageLocationId)
projectDestination$projectId <- folderId

projectDestination <- synRestPOST('/projectSettings', body=toJSON(projectDestination))
```

> You can then upload files to this STS Storage Location using the normal methods, either through [the Synapse website](https://www.synapse.org/) or through the client of your choice.

Reviewer 4: I see how it is possible to add FileHandles for files uploaded via STS tokens (or otherwise) to an external bucket using the **POST /externalFileHandle/s3** endpoint. However for Synapse storage (i.e. non-external), after getting an STS token and using it to upload a file directly to a Synapse managed S3 bucket I don't see an equivalent way to be able to add a FileHandle representation for that uploaded file. The externalFileHandle/s3 endpoint doesn't work properly rejecting it as a non external storage location.

Also, per @thomasyu888 's point about the accessibility of the bucket name, to upload a file at all to Synapse storage using an STS token and an external S3 client one must already know the bucket name as it doesn't seem to be accessible via the upload destination or storage location.

Author: Good catch. I've filed https://sagebionetworks.jira.com/browse/PLFM-6227 to address this.

## Custom STS Storage Locations

You can also create an STS Storage Location using a custom storage location in an external AWS S3 bucket. To set this up, first follow the steps in [Custom Storage Locations](custom_storage_location.md), up until the step "Set S3 Bucket as Upload Location". Instead, make sure you have an empty Synapse folder, then run the following code.

> {% include important.html content="You should specify a baseKey in your storage location, which is a folder path in your bucket that all files in the storage location should go into. The baseKey is optional, but if it is not specified, the temporary AWS credentials vended by STS will give users access to the whole bucket." %}

Reviewer 3: "...will give users access to the whole bucket".  Is this really true or is it instead that the storage location has a root key prefix like <userid>/<uuid> that the STS token is constrained to?

Author: This is true. If the baseKey is not specified, STS grants access to the whole bucket.

We can prevent this by auto-generating a baseKey, but this seems needlessly restrictive as users may *want* to grant access to the whole bucket.

Reviewer 5: I amended the language to be stronger here, since this sounds like it could be a very undesired outcome.

> {% include important.html content="You should specify a baseKey in your storage location, which is a folder path in your bucket that all files in the storage location should go into. The baseKey is optional, but if it is not specified, the temporary AWS credentials vended by STS will give users access to the whole bucket." %}

Reviewer 5: ```suggestion
{% include important.html content="If a baseKey is **not** specified, the temporary AWS credentials vended by STS will give users access to the whole bucket. To prevent access to the whole bucket, enter a folder path in your bucket that all files in the storage location should go into as the baseKey. " %}
```

{% include note.html content="We recommend picking one of these methods and sticking to it and avoiding mixing methods." %}

## Obtaining Temporary S3 Credentials

Once your STS Storage Location is set up, you can use Synapse to request temporary AWS credentials with direct access to your data in S3. These temporary credentials are good for up to 12 hours.

To get these temporary credentials, you may use our [REST interface](https://rest-docs.synapse.org/rest/GET/entity/id/sts.html).

Alternatively, you may use the below sample Java code:

```java
StsCredentials stsCredentials = synapseClient.getTemporaryCredentialsForEntity(folderEntityId, StsPermission.read_only);
AWSCredentials awsCredentials = new BasicSessionCredentials(stsCredentials.getAccessKeyId(), stsCredentials.getSecretAccessKey(), stsCredentials.getSessionToken());
AWSCredentialsProvider awsCredentialsProvider = new AWSStaticCredentialsProvider(awsCredentials);

> AmazonS3 s3Client = AmazonS3ClientBuilder.standard().withCredentials(awsCredentialsProvider).build();

Reviewer: Python equivalent here:

```

import synapseclients
import boto3

syn = synapseclient.login()
sts_credentials = syn.restGET(f"/entity/{FOLDER}/sts?permission=read_only")
client = boto3.client(
    's3',
    aws_access_key_id=sts_credentials['accessKeyId'],
    aws_secret_access_key=sts_credentials['secretAccessKey'],
    aws_session_token=sts_credentials['sessionToken'],
)
client.download_file('mybucket', 'hello.txt', '/tmp/hello.txt')
# mybucket -- The name of the bucket to download from.
# hello.txt -- The name of the key to download from.
# /tmp/hello.txt -- The path to the file to download to.

```

But I would like to provide a working example to actually download the file

Reviewer 5: @thomasyu888 can this code example live in https://github.com/Sage-Bionetworks/synapsePythonClient/pull/746/files#?

---
title: "AWS Security Token Service Storage Locations"
layout: article
excerpt: Follow these steps to set up STS Storage Locations and use them to access your S3 data directly.
category: admin-and-settings
---

Using AWS Security Token Service (STS), Synapse can securely grant you temporary AWS credentials with access to data directly in S3. This can be useful if you want to:

> * Transfer data in bulk

Author: Yes. I think bulk download is still one of the major scenarios here, and bulk download is supported for both Synapse storage and external S3 storage.

* Allow your compute cluster to read S3 objects using the S3 APIs

All of which you can now do with minimal overhead from Synapse.

> {% include note.html content="You can only create STS Storage Locations on an empty folder. This is to ensure consistency between the STS Storage Location and the Synapse folder." %}

Reviewer 5: The feature does not address this use case - https://sagebionetworks.jira.com/browse/PLFM-6282

                      'settingsType': 'upload'}
project_destination['locations'] = [destination['storageLocationId']]
project_destination['projectId'] = FOLDER

project_destination = syn.restPOST('/projectSettings', body = json.dumps(project_destination))
```

##### R

```r
#set storage location
library(synapser)
library(rjson)
synLogin()
folderId <- 'syn12345'

destination <- list(uploadType='S3',
                    stsEnabled=TRUE,
                    concreteType='org.sagebionetworks.repo.model.project.S3StorageLocationSetting')
destination <- synRestPOST('/storageLocation', body=toJSON(destination))

projectDestination <- list(concreteType='org.sagebionetworks.repo.model.project.UploadDestinationListSetting',
                           settingsType='upload')
projectDestination$locations <- list(destination$storageLocationId)
projectDestination$projectId <- folderId

projectDestination <- synRestPOST('/projectSettings', body=toJSON(projectDestination))
```

You can then upload files to this STS Storage Location using the normal methods, either through [the Synapse website](https://www.synapse.org/) or through the client of your choice.

> ## Custom STS Storage Locations

Author: "Custom" is what the external storage location page calls them, but internally, we call them external storage locations.

                    baseKey='nameofyourbasekey',
                    concreteType='org.sagebionetworks.repo.model.project.ExternalS3StorageLocationSetting')
destination <- synRestPOST('/storageLocation', body=toJSON(destination))

projectDestination <- list(concreteType='org.sagebionetworks.repo.model.project.UploadDestinationListSetting',
                           settingsType='upload')
projectDestination$locations <- list(destination$storageLocationId)
projectDestination$projectId <- folderId

projectDestination <- synRestPOST('/projectSettings', body=toJSON(projectDestination))
```

Once your STS Storage Location is set up on your Synapse folder, you can add files to Synapse in one of two ways:

1. You can upload files through Synapse, either through [the Synapse website](https://www.synapse.org/) or through the client of your choice.
2. If you plan to upload files directly to your S3 bucket, or if you already have files in your S3 bucket, you can add representations of those files to Synapse programmatically. Follow the steps at [Adding Files in Your S3 Bucket to Synapse](custom_storage_location.md#adding-files-in-your-s3-bucket-to-synapse).

> {% include note.html content="We recommend picking one of these methods and sticking to it and avoiding mixing methods." %}

Author: I updated the comment with further explanation: "Synapse automatically generates folders for files uploaded through the Synapse website or clients. Files added directly to S3 may not match this folder structure. We recommend against mixing these two methods of adding files to prevent confusion in the folder structure."

## Obtaining Temporary S3 Credentials

Once your STS Storage Location is set up, you can use Synapse to request temporary AWS credentials with direct access to your data in S3. These temporary credentials are good for up to 12 hours.

To get these temporary credentials, you may use our [REST interface](https://rest-docs.synapse.org/rest/GET/entity/id/sts.html).

{% include note.html content="If your storage location is external storage, you may request read-only or read-write permissions. If that storage location is Synapse storage, you must request read-only permissions." %}

Alternatively, you may use the below sample Java code:

```java
StsCredentials stsCredentials = synapseClient.getTemporaryCredentialsForEntity(folderEntityId, StsPermission.read_only);
AWSCredentials awsCredentials = new BasicSessionCredentials(stsCredentials.getAccessKeyId(), stsCredentials.getSecretAccessKey(), stsCredentials.getSessionToken());
AWSCredentialsProvider awsCredentialsProvider = new AWSStaticCredentialsProvider(awsCredentials);
AmazonS3 s3Client = AmazonS3ClientBuilder.standard().withCredentials(awsCredentialsProvider).build();
```

## Migrating Existing Data

### From an Existing S3 Bucket

> If you have existing data in an S3 bucket, either standalone data or data from a previous Synapse project, you can use [this sample code](https://github.com/Sage-Bionetworks/Synapse-Repository-Services/blob/develop/client/sample-code/src/main/java/org/sagebionetworks/sample/sts/MigrateS3Bucket.java) to migrate your S3 data to a new Synapse folder with an STS Storage Location.

Author: This particular sample code is for populating an External Storage Location with an existing S3 bucket. I wrote it in Java because I don't know R or Python.
